### ISSUE TITLE
NFS volume plugin should translate to service IP before making mount call

### ISSUE THREAD

**Issue by zmerlynn** [kubernetes/kubernetes#15995]
I'm trying to brush off the (nfs example)[https://github.com/kubernetes/kubernetes/tree/master/examples/nfs] now that privileged is working by default, and one of the small warts in it is the web server pod needs a hardcoded IP or hostname in order to bring the pieces together. (And I'm running through it for GCE to make sure it works there.)

I think this is straightforward to fix with an API call during the mount, then we can just call system mount with the cluster IP.

cc @thockin @cjcullen @jsafrane

**Comment by thockin** [comment]
I could see adding an EndpointsName alongside Server and using one or the
other,  Unfortunately Server is not marked as optional, so we'll have to
figure out the compat.

**Comment by zmerlynn** [comment]
Yeah, having a small API modification would force the disambiguation of the service. My initial thought was the ghetto "does this look DNS-like?", which is going to be too brittle across API compat.

I'm not seeing a big compat issue. I think the EndpointsName carrying volumes would need translating to Server inside the volume plugin (though extend it to Server:Port, since the NFS mount command allows that and endpoints will certainly carry it), then the code paths are common. I'm probably being naive somewhere, though.

**Comment by thockin** [comment]
Our API documents Server as a required field.  This means a consumer can
safely assume it is present and non-empty and that it is the canonical
representation of the NFS server.  Changing it to optional potentially
breaks those clients.

That said, this is PROBABLY pretty low impact, but it's a bad precedent.
We're trying to be better than that.

**Comment by zmerlynn** [comment]
@thockin: Now I understand, thanks.

**Comment by elibixby** [comment]
Bumping this issue as it is also something I would like. Not a super important use case, but having this feature would make it easier to make click-to-deploy style deployments of NFS in k8s with e.g. helm. 

As an API change option why not add an optional `usesDNS` or such field, which signals that the `server` field is a dns address and not an IP. This preserves `server` as a required field and avoids kludgy parsing to determine whether the API call should be made or not.

**Comment by thockin** [comment]
I'm trying to page in the details of this bug but something eludes me.  The docks for NFS say "Server is the hostname or IP address of the NFS server".  So what's not working?

**Comment by elibixby** [comment]
You can't use cluster DNS as the hostname. It would make it really easy to
set up a NFS server behind a PV (i.e. one kubectl create -f ) if you
could...

**Comment by thockin** [comment]
Ahh, right, because the host doesn't use kube-DNS.  Thanks for the kick in
the head.  We might be able to get away with a `resolveAsServiceName`
boolean to mutate processing of the Server name.  Does someone have time to
prototype?

**Comment by bprashanth** [comment]
Isn't it easiest to just lookup the ip from full service name in the field, from the plugin?
This would also help petset clustering by making it easier to share a temp nfs volume during bootstrap.

**Comment by JeanMertz** [comment]
I just came across the same issue. This would be a nice improvement if this was made possible 👍
